docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+provisioner-specific.
+
+.. code:: JSON
+
+      "metadata": {
+        "kernel_provisioner": {
+          "provisioner_name": "k8s-provisioner",
+          "config": {
+              "image_name": "my_docker_org/kernel:2.1.5",
+              "max_cpus": 4
+          }
+        }
+      },
+
+Kernel provisioner authors implement their provisioners by deriving from
+:class:`KernelProvisionerBase` and expose their provisioner for consumption
+via entry-points:
+
+.. code:: python
+
+    'jupyter_client.kernel_provisioners': [
+                'k8s-provisioner = my_package:K8sProvisioner',
+            ],
+
+Backwards Compatibility
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Prior to this release, no ``kernel.json`` (kernelspec) will contain a
+provisioner entry, yet the framework is now based on using provisioners.
+As a result, when a ``kernel_provisioner`` stanza is **not** present in
+a selected kernelspec, jupyter client will, by default, use the built-in
+``LocalProvisioner`` implementation as its provisioner. This provisioner
+retains today's local kernel functionality. It can also be subclassed
+for those provisioner authors wanting to extend the functionality of
+local kernels. The result of launching a kernel in this manner is
+equivalent to the following stanza existing in the ``kernel.json`` file:
+
+.. code:: JSON
+
+      "metadata": {
+        "kernel_provisioner": {
+          "provisioner_name": "local-provisioner",
+          "config": {
+          }
+        }
+      },
+
+Should a given installation wish to use a *different* provisioner as
+their "default provisioner" (including subclasses of
+``LocalProvisioner``), they can do so by specifying a value for
+``KernelProvisionerFactory.default_provisioner_name``.
+
+Implementing a custom provisioner
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The impact of Kernel Provisioning is that it enables the ability to
+implement custom kernel provisioners to manage a kernel's lifecycle
+within any runtime environment. There are currently two approaches by
+which that can be accomplished, extending the ``KernelProvisionerBase``
+class or extending the built-in class - ``LocalProvisioner``. As more
+provisioners are introduced, some may be implemented in an abstract
+sense, from which specific implementations can be authored.
+
+Extending ``LocalProvisioner``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you're interested in running kernels locally and yet adjust their
+behavior, there's a good chance you can simply extend
+``LocalProvisioner`` via subclassing. This amounts to deriving from
+``LocalProvisioner`` and overriding appropriate methods to provide your
+custom functionality.
+
+In this example, RBACProvisioner will verify whether the current user is
+in the role meant for this kernel. If not, an exception will be thrown.
+
+.. code:: python
+
+    class RBACProvisioner(LocalProvisioner):
+
+        role: str = Unicode(config=True)
+
+        async def pre_launch(self, **kwargs: Any) -> Dict[str, Any]:
+
```

Comment:
Looks like there is a well-defined lifecycle, cfr `pre_launch`. Is it possible to list those lifecycle methods here (or link them to the API docs, or to the source)?

Reply:
I'm not sure what you mean by this - nearly every method pertains to the lifecycle of the kernel processing since that is exactly what `KernelManager` (which drives the calls on the provisioner) is responsible for.  Are you suggesting that we document in the provisioner's method where/when they are called from the `KernelManager`?

Could you please provide an example?  I'm happy to add that kind of thing provided it adds value.

I do add the call relationship in the [base provisioner class](https://github.com/jupyter/jupyter_client/pull/612/files#diff-b370bb301ffe7b9de4edb6f77835f50f930c12552fdacd203839f9710daeedb5R144-R161) - which I view as the canonical doc text for all provisioners.

Reply:
> nearly every method pertains to the lifecycle of the kernel processing since that is exactly what KernelManager

Then it is fine to not embed more details in the doc.

> Are you suggesting that we document in the provisioner's method where/when they are called from the KernelManager?

I did some time ago a sequence diagram for the kernel providers (not merged). We can add that kind of details later on if it makes sense.

Reply:
Yes, that was helpful to the providers' discussion and a provisioner-based version would be helpful as well.